core/arch/arm/kernel/generic_boot.c

```diff
@@ -179,7 +180,120 @@ static void init_vfp_sec(void)
 }
 #endif
 
+#ifdef CFG_CORE_SANITIZE_KADDRESS
+static void init_run_constructors(void)
+{
+	const vaddr_t *ctor;
+
+	for (ctor = &__ctor_list; ctor < &__ctor_end; ctor++)
+		((void (*)(void))(*ctor))();
+}
+
+static void init_asan(void)
+{
+
+	/*
+	 * CFG_ASAN_SHADOW_OFFSET is also supplied as
+	 * -fasan-shadow-offset=$(CFG_ASAN_SHADOW_OFFSET) to the compiler.
+	 * Since all the needed values to calculate the value of
+	 * CFG_ASAN_SHADOW_OFFSET isn't available in to make we need to
+	 * calculate it in advance and hard code it into the platform
+	 * conf.mk. Here where we have all the needed values we double
+	 * check that the compiler is supplied the correct value.
+	 */
+
+#define __ASAN_SHADOW_START \
+	ROUNDUP(TEE_RAM_VA_START + (CFG_TEE_RAM_VA_SIZE * 8) / 9 - 8, 8)
+	assert(__ASAN_SHADOW_START == (vaddr_t)&__asan_shadow_start);
+#define __CFG_ASAN_SHADOW_OFFSET \
+	(__ASAN_SHADOW_START - (TEE_RAM_VA_START / 8))
+	COMPILE_TIME_ASSERT(CFG_ASAN_SHADOW_OFFSET == __CFG_ASAN_SHADOW_OFFSET);
+#undef __ASAN_SHADOW_START
+#undef __CFG_ASAN_SHADOW_OFFSET
+
+	/*
+	 * Assign area covered by the shadow area, everything from start up
+	 * to the beginning of the shadow area.
+	 */
+	asan_set_shadowed((void *)TEE_TEXT_VA_START, &__asan_shadow_start);
+
+	/*
+	 * Add access to areas that aren't opened automatically by a
+	 * constructor.
+	 */
+	asan_tag_access(&__initcall_start, &__initcall_end);
+	asan_tag_access(&__ctor_list, &__ctor_end);
+	asan_tag_access(__rodata_start, __rodata_end);
+#ifdef CFG_WITH_PAGER
+	asan_tag_access(__pageable_start, __pageable_end);
+#endif /*CFG_WITH_PAGER*/
+	asan_tag_access(__nozi_start, __nozi_end);
+
+	init_run_constructors();
+
+	/* Everything is tagged correctly, let's start address sanitizing. */
+	asan_start();
+}
+#else /*CFG_CORE_SANITIZE_KADDRESS*/
+static void init_asan(void)
+{
+}
+#endif /*CFG_CORE_SANITIZE_KADDRESS*/
+
 #ifdef CFG_WITH_PAGER
+
+#ifdef CFG_CORE_SANITIZE_KADDRESS
+static void carve_out_asan_mem(tee_mm_pool_t *pool)
+{
+	const size_t s = pool->hi - pool->lo;
+	tee_mm_entry_t *mm;
+	paddr_t apa = ASAN_MAP_PA;
+	size_t asz = ASAN_MAP_SZ;
+
+	if (core_is_buffer_outside(apa, asz, pool->lo, s))
+		return;
+
+	/* Reserve the shadow area */
+	if (!core_is_buffer_inside(asz, asz, pool->lo, s)) {
+		if (apa < pool->lo) {
+			/*
+			 * ASAN buffer is overlapping with the beginning of
+			 * the pool.
+			 */
+			asz -= pool->lo - apa;
+			apa = pool->lo;
+		} else {
+			/*
+			 * ASAN buffer is overlapping with the end of the
+			 * pool.
+			 */
+			asz = apa - pool->hi;
+		}
+	}
```

Comment:
do we need to assert asan map is not fully overlapping the target pool ?   
`assert(apa >= pool.lo || (apa + asz) <= pool->hi);`
(unlikely to happen)